upgrade/upgrades: TestUpgradeSchemaChangerElements failed

upgrade/upgrades.TestUpgradeSchemaChangerElements [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8940378?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8940378?buildTab=artifacts#/) on master @ [1d4feea096e9714b083cca5cda8b89706ef9e35b](https://github.com/cockroachdb/cockroach/commits/1d4feea096e9714b083cca5cda8b89706ef9e35b):

Fatal error:

```
panic: test timed out after 14m55s

```
Stack: 

```
goroutine 293683 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x88
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestUpgradeSchemaChangerElements
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/3538dab3e8bcdb43d30e6320ed019d4d/logTestUpgradeSchemaChangerElements1322792266
    test_log_scope.go:79: use -show-logs to present logs inline
=== RUN   TestUpgradeSchemaChangerElements/running_schema_change_that_will_have_a_deprecated_waitForSchemaChangerElementMigration_element
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-schema
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestUpgradeSchemaChangerElements.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-25059

upgrade/upgrades.TestUpgradeSchemaChangerElements [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8970048?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8970048?buildTab=artifacts#/) on master @ [1b162d1b274eec7b307fbbfca7294460bfdef025](https://github.com/cockroachdb/cockroach/commits/1b162d1b274eec7b307fbbfca7294460bfdef025):

Fatal error:

```
panic: test timed out after 14m55s

```
Stack: 

```
goroutine 338193 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x88
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestUpgradeSchemaChangerElements
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/9a9dbc5a7e380eaf5ef4ce24e3fe45dd/logTestUpgradeSchemaChangerElements3200980430
    test_log_scope.go:79: use -show-logs to present logs inline
=== RUN   TestUpgradeSchemaChangerElements/running_schema_change_that_will_have_a_deprecated_waitForSchemaChangerElementMigration_element
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestUpgradeSchemaChangerElements.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/upgrade/upgrades/schemachanger_elements_test.go
+++ b/pkg/upgrade/upgrades/schemachanger_elements_test.go
@@ -27,7 +27,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/tests"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
-	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/util"
 	"github.com/cockroachdb/cockroach/pkg/util/ctxgroup"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
@@ -37,7 +36,6 @@ import (
 
 func TestUpgradeSchemaChangerElements(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	skip.WithIssue(t, 98062, "flaky test")
 	defer log.Scope(t).Close(t)
 
 	ctx := context.Background()
@@ -120,6 +118,7 @@ func TestUpgradeSchemaChangerElements(t *testing.T) {
 				schemaChangeAllowedToComplete chan struct{}
 				waitedForJob                  chan struct{}
 				jobIsPaused                   chan struct{}
+				jobStarted                    chan struct{}
 				readyToQuery                  int64 = 0
 			)
 			scJobID := int64(jobspb.InvalidJobID)
@@ -131,6 +130,9 @@ func TestUpgradeSchemaChangerElements(t *testing.T) {
 					if p.Stages[stageIdx].Phase == scop.PreCommitPhase {
 						atomic.StoreInt64(&scJobID, int64(p.JobID))
 					} else if p.Stages[stageIdx].Phase > scop.PreCommitPhase {
+						if tc.unpauseJob {
+							jobStarted <- struct{}{}
+						}
 						<-waitedForJob
 						waitedForJob = nil
 						schemaChangeAllowedToComplete <- struct{}{}
@@ -156,19 +158,23 @@ func TestUpgradeSchemaChangerElements(t *testing.T) {
 			params.Knobs.JobsTestingKnobs = jobKnobs
 			s, sqlDB, _ := serverutils.StartServer(t, params)
 			defer s.Stopper().Stop(ctx)
-			_, err := sqlDB.Exec("CREATE TABLE t (pk INT PRIMARY KEY, b INT)")
+			// Make sure our cluster is up
+			_, err := sqlDB.Exec("SET CLUSTER SETTING version=$1",
+				clusterversion.ByKey(clusterversion.V23_1_SchemaChangerDeprecatedIndexPredicates-1).String())
+			require.NoError(t, err)
+			_, err = sqlDB.Exec("CREATE TABLE t (pk INT PRIMARY KEY, b INT)")
 			require.NoError(t, err)
 			_, err = sqlDB.Exec("CREATE UNIQUE INDEX ON t (b) WHERE pk > 0")
 			require.NoError(t, err)
 			_, err = sqlDB.Exec("SET CLUSTER SETTING jobs.debug.pausepoints='newschemachanger.before.exec'")
 			require.NoError(t, err)
-
 			ctx := context.Background()
 			g := ctxgroup.WithContext(ctx)
 
 			schemaChangeAllowedToComplete = make(chan struct{})
 			waitedForJob = make(chan struct{})
 			jobIsPaused = make(chan struct{})
+			jobStarted = make(chan struct{})
 			g.GoCtx(func(ctx context.Context) error {
 				err := tc.run(t, sqlDB)
 				jobIsPaused <- struct{}{}
@@ -183,8 +189,14 @@ func TestUpgradeSchemaChangerElements(t *testing.T) {
 					require.NoError(t, err)
 					_, err = sqlDB.Exec("RESUME JOB $1", jobspb.JobID(atomic.LoadInt64(&scJobID)))
 					require.NoError(t, err)
+					// Wait for the job to start after the unpause,
+					// we will let it get stuck during execution next.
+					// We don't want a race condition with the migration
+					// job that will pause it again.
+					<-jobStarted
 				}
-				_, err = sqlDB.Exec("SET CLUSTER SETTING version = crdb_internal.node_executable_version()")
+				_, err = sqlDB.Exec("SET CLUSTER SETTING version = $1",
+					clusterversion.ByKey(clusterversion.V23_1_SchemaChangerDeprecatedIndexPredicates).String())
 				return err
 			})
 

